csrc/abstract_tensor.cpp

```diff
@@ -413,4 +413,41 @@ std::vector<AbstractTensor> AbstractTensor::unzip() const {
   return result;
 }
 
+namespace {
+
+struct DispatchParallelize {
+  template <typename INPUT>
+  void operator()(ParallelType parallel_type, INPUT&& in) const {
+    using IN = std::decay_t<INPUT>;
+    if constexpr (std::is_same_v<IN, std::monostate>) {
+      return;
+    } else if constexpr (std::is_same_v<IN, IterDomain*>) {
+      return in->parallelize(parallel_type);
+    } else if constexpr (std::is_same_v<IN, ValGroupAndItsGraph>) {
+      // TODO: instead of throwing an error, should we just parallelize all IDs
+      // in the ValGroup? This behavior only makes sense for the loop graph, so
+      // we are disabling it for now to avoid unexpected behavior. If in the
+      // future, it does turn out that we need to commonly manually construct a
+      // loop graph, we can turn this on.
+      NVF_ERROR(false, "Parallelizing ValGroup is not supported");
```

Comment:
Regarding the TODO: Since we might not have "replayed" all the tensors yet, I think it might make more sense to keep parallelization type with the `ValGroup` so that we can apply it to a tensor later after replaying.

Reply:
What do you mean by "keep parallelization type with the `ValGroup`"? Something like keeping a `std::unordered_map<ValGroup, ParallelType>`?

Reply:
Yeah, something like that, or `ValGroupWithItsGraphAndParallelType` :smile: .